### PR TITLE
Tests now have a seeded random state, for reproducibility (closes #958)

### LINKED ISSUE #958: Stabilize coverage fluctuations due to random numbers in tests.
As @richardjgowers pointed out in #955, coverage is currently reported with fluctuations of around 0.1% because the use of random numbers causes slightly different code paths to be followed.

One might argue that randomness should be avoided in tests, but it's a handy way to quickly generate on-the-fly test objects from scratch (like sets of coordinates).

I came up with a simple test plugin that'll reseed the `random` and `np.random` RNGs prior to each test, so that we get consistent creation of test values, and a stable code path. In turn, this will allow us to be more stringent when thresholding coverage drops (#955). I guess it'll be one more thing to convert when moving away from nose tests, but it's also not that critical anyway.

Perhaps the cleanest alternative is to hardcode those "random" test values, either as test datafiles or in the test scripts themselves. The fact that we follow different code paths dependent on input suggests we will want to control that at some point to be able to get full coverage. This, of course, will take much more work to solve.

### PR DESCRIPTION
**Pull request by mnmelo**
Use of the new `RandomState` class is documented in its docstring, in `MDAnalysisTests.rng.RandomState`.

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

MDAnalysisTests.rng.RandomState test base class creates a random and a
numpy.random state per test instance.

Adjusted all occurrences where consistent state is needed.

Fixed test_imports, which didn't check multiple directory levels.

Minor docstring cleanup in MDAnalysis/lib/transformations.py

### CODE REVIEW

**Comment by richardjgowers** [comment]
So just so I understand, this means all tests use the same random numbers from now on? If so, looks good.

**Comment by mnmelo** [comment]
Yup, as long as they use `self.np_random` or `self.random`.

I did it with a base class so that you don't have to explicitly instantiate the random state per test, although it boils down to pretty much the same. Let me know if you'd rather have a different design.

Also, when we later want to set the seed for specific tests it's as easy as doing `self.np_random.seed(new_seed)`.

As far as reproducibility goes, repeated tests in parallel gave me the exact same coverage. A quirk, though: in parallel some files seem to be skipped when accounting for coverage (`core/transformations.py`, for instance). I'll double-check, but don't believe it's related to these changes.

**Comment by mnmelo** [comment]
Well, in coveralls.io `core/transformations.py` shows up as fully covered, contrary to my local run. I won't worry about it, then.

File: testsuite/MDAnalysisTests/rng.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
+#
+
+import numpy as np
+from numpy.testing import TestCase
+import random
+
+seed = 3856349111 # from: np.uint32(hash("MDAnalysis"))
+
+class RandomState(TestCase):
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/rng.py]: Rather than have this as a class mixin, could we not just have np_random as a function in the MdaTests.rng module, then import it to use it?

Comment (mnmelo) [inline testsuite/MDAnalysisTests/rng.py]: Yea, I don't love the mixin either, but with the function call approach, where is `np.random.RandomState` initialized for each test? Do we do it by hand? Then we don't even need a function call and just use the `np.random.RandomState` object directly (in this case the `setUp` of each test that needs random state gets messier).

Comment (kain88-de) [inline testsuite/MDAnalysisTests/rng.py]: You can use a function with a default seed. That can return a RandomState. Such a function will be very easy to convert into a fixture in pytest.

Comment (mnmelo) [inline testsuite/MDAnalysisTests/rng.py]: Or, in slightly hackish way, `MDAnalysisTests.rng` could store the `np.random.RandomState` objects in a module-level dict hashed by the test's class instance. This would emulate class-based caching, without the need for explicitly inheriting anything.

Comment (richardjgowers) [inline testsuite/MDAnalysisTests/rng.py]: I think Max is right, just make a function that does the current setUp
code, then calls random and returns that

Comment (kain88-de) [inline testsuite/MDAnalysisTests/rng.py]: > Or, in slightly hackish way, MDAnalysisTests.rng could store the np.random.RandomState objects in a module-level dict hashed by the test's class instance. This would emulate class-based caching, without the need for explicitly inheriting anything.

This sounds way to complicated for what we want to do. Why would we need caching anyway?

Comment (mnmelo) [inline testsuite/MDAnalysisTests/rng.py]: @kain88-de's way only works for the simplest cases. Many tests call random numbers several times (say, to make two random matrices). If said function initializes the state every time the two matrices will always be the same. Unless you want to pass to the test the burden of specifying a new seed every time you call random.

In addition, some tests call specialized methods, such as `np.random.normal`. It isn't clear how to deal with these if all you have is a function call.

What if I change the current approach from "Is a RandomState" to a "Has a RandomState" in a way that tests have an instance of `MDAnalysisTests.rng.RandomState`? (so no need for inheriting anything, and we avoid having to call `super`).

Comment (kain88-de) [inline testsuite/MDAnalysisTests/rng.py]: what is so wrong with 

```python
class TestSomething:
    def setUp(self):
        self.rng = mda.testing.rng()
    
    def test_foo(self):
        random = self.rng.normal()
```

Comment (mnmelo) [inline testsuite/MDAnalysisTests/rng.py]: Ah, I get what you mean now. I misread your suggestion as a function that returns a random values. I'll adapt the code to this.

One question: we sometimes draw random numbers from both `numpy.random` and `random`. For simplicity, shall I change all cases, where state is needed, to use `numpy.random`? This way we only need to initialize one type of state.

Comment (kain88-de) [inline testsuite/MDAnalysisTests/rng.py]: Yes that would be good. There isn't a functional difference between the two as they all use the mersenne twister engine. I assume that neither version has a serious bug in the distributions they generate.

**Comment by mnmelo** [comment]
Ok guys, I'm trying the new idea, to have a function return a `RandomState`. But I'm bumping into a lot of duplicated code when initializing the states: many test classes use random numbers, but only a few define `setUp` methods. I'm having to add a lot of these `setUp` methods just to initialize the `RandomState`, which seems very redundant.

The alternative is to initialize the `RandomState` inside every `test_*` method, which also quickly becomes cumbersome.

Opinions?

**Comment by richardjgowers** [comment]
The more I think about it we shouldn't be using random numbers at all - the tests should be identical each time. I'd rather just replace all random numbers with some prerolled random numbers hardcoded in, maybe even a loop over `np.linspace` to cover all corner cases.  I don't mind getting this done tomorrow.

**Comment by mnmelo** [comment]
I guess that might be the best solution, `linspace` (+ reshape, where needed). It's already used as "random" input in some other tests.

**Comment by mnmelo** [comment]
Actually, there's really no difference between a seeded RNG and a hand made array. But the latter will be more cumbersome to create. I still vote for the RNG approach.

File: testsuite/MDAnalysisTests/test_imports.py [hunk 2]
```diff
@@ -34,6 +40,11 @@ def _run_test_relative_import(testing_module):
                         "module {testing_module} at linenumber {lineno}.".format(**vars()))
 
     def test_relative_imports(self):
-        list_testing_modules = glob.glob(os.path.join(self.path_to_testing_modules, '*.py'))
-        for testing_module in list_testing_modules:
-            yield self._run_test_relative_import, testing_module
+        for dirpath, dirnames, files in os.walk(self.path_to_testing_modules):
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/test_imports.py]: This should probably get span out into a new PR if this is fixing something? @mnmelo

Comment (mnmelo) [inline testsuite/MDAnalysisTests/test_imports.py]: Good point.
